src/common/selftest/selftest_esp_update.cpp

```diff
@@ -0,0 +1,344 @@
+            uint8_t buffer[buffer_length];
+
+            readBytes = fread(buffer, 1, sizeof(buffer), opened_file);
+            readCount += readBytes;
+            log_debug(Network, "ESP read data %ld", readCount);
+            if (ferror(opened_file)) {
+                log_error(Network, "ESP flash: Unable to read file %s", current_file->filename);
+                readBytes = 0;
+                progress_state = esp_upload_action::USB_error;
+                break;
+            }
+            if (readBytes > 0) {
+                if (esp_loader_flash_write(buffer, readBytes) != ESP_LOADER_SUCCESS) {
+                    log_error(Network, "ESP flash write FAIL");
+                    progress_state = esp_upload_action::ESP_error;
+                    break;
+                } else {
+                    updateProgress();
+                }
+            } else {
+                fclose(opened_file);
+                ++current_file;
+                progress_state = current_file != firmware_set.end()
+                    ? esp_upload_action::Start_flash
+                    : esp_upload_action::Finish_and_reset;
+            }
+            break;
+        }
+        case esp_upload_action::Finish_and_reset:
+            log_info(Network, "ESP finished flashing");
+            phase = PhasesSelftest::ESP_passed;
+            esp_loader_flash_finish(true);
+            espif_flash_deinitialize();
+            notify_reconfigure();
+            progress_state = esp_upload_action::Finish_wait_user;
+            current_file = firmware_set.begin();
+            readCount = 0;
+            break;
+        case esp_upload_action::Finish_wait_user:
+            if (continue_pressed) {
+                progress_state = esp_upload_action::Done;
+            }
+            break;
+        case esp_upload_action::ESP_error:
+        case esp_upload_action::USB_error: {
+            phase = PhasesSelftest::ESP_failed;
+            esp_loader_flash_finish(false);
+            progress_state = esp_upload_action::Error_wait_user;
+            current_file = firmware_set.begin();
+            readCount = 0;
+            fclose(opened_file);
+            break;
+        }
+        case esp_upload_action::Error_wait_user:
+            if (continue_pressed) {
+                progress_state = esp_upload_action::Done;
+            }
+            break;
+        case esp_upload_action::Aborted:
+            progress_state = esp_upload_action::Done;
+            break;
+        case esp_upload_action::Done:
+            break;
+        }
+
+        //actualize status
+        status_encoder_union u;
+        u.status.phase = phase;
+        u.status.progress = progress;
+        u.status.current_file = current_file_no;
+        u.status.count_of_files = files_to_upload;
+        status = u.u32;
+    }
+}
+
+static void EspTask(void *pvParameters) {
+    ESPUpdate update((int)pvParameters);
+
+    update.Loop();
+
+    vTaskDelete(NULL); // kill itself
```

Comment:
Is a task allowed to commit suicide in such a way?

Reply:
it is from freertos sample code

Reply:
ok